server/test/integration/dbTest/beginTransactionTest.js

```diff
@@ -0,0 +1,112 @@
+  })
+
+  describe('when beginTransaction fails the first time because of a deadlock, then successfully connects after retry', () => {
+    let clientStub
+    beforeEach(async () => {
+      clientStub = { query: sandbox.stub() }
+      sandbox.spy(db, 'beginTransaction')
+      poolConnectStub.onCall(0).rejects(new Error('deadlock detected')).onCall(1).resolves(clientStub)
+      pgClient = await db.beginTransaction()
+    })
+
+    it('should log the deadlock error', () => {
+      expect(helpers.logError).to.be.calledWith(`Error running the beginTransaction query: Error: deadlock detected`)
+    })
+
+    it('should log the retry', () => {
+      expect(helpers.log).to.be.calledWith(`Retrying beginTransaction.`)
+    })
+
+    it('should lock the clients, sessions, and locations tables to reflect the retry', () => {
+      expect(clientStub.query).to.be.calledWith(`LOCK TABLE clients, sessions, locations`)
+    })
+
+    it('should return a valid pgClient', () => {
+      expect(pgClient).to.be.equal(clientStub)
+    })
+  })
+
+  describe('when beginTransaction is successful the first time it is called', () => {
+    let clientStub
+    beforeEach(async () => {
+      clientStub = { query: sinon.stub() }
+      sandbox.spy(db, 'beginTransaction')
+      poolConnectStub.onCall(0).resolves(clientStub)
+      pgClient = await db.beginTransaction()
+    })
+
+    it('should not log any errors', () => {
+      expect(helpers.logError).to.not.be.called
+    })
+
+    it('should not log a retry', () => {
+      expect(helpers.log).to.not.be.calledWith(`Retrying beginTransaction.`)
+    })
+
+    it('should lock the clients, sessions, and locations tables', () => {
+      expect(clientStub.query).to.be.calledWith(`LOCK TABLE clients, sessions, locations`)
+    })
+
+    it('should return a valid pgClient', () => {
+      expect(pgClient).to.be.equal(clientStub)
+    })
+  })
+
+  describe('when beginTransaction fails twice, deadlocking on the first attempt and throwing an error with a valid pgClient on the second attempt', () => {
+    let clientStub
+    beforeEach(async () => {
+      clientStub = { query: sandbox.stub() }
+      clientStub.query.withArgs('BEGIN').rejects(new Error('some error'))
+      sandbox.spy(db, 'beginTransaction')
+      poolConnectStub.onCall(0).rejects(new Error('deadlock detected')).onCall(1).resolves(clientStub)
+      pgClient = await db.beginTransaction()
+    })
+
+    it('should log the deadlock error', () => {
+      expect(helpers.logError).to.be.calledWith(`Error running the beginTransaction query: Error: deadlock detected`)
+    })
+
+    it('should log the retry', () => {
+      expect(helpers.log).to.be.calledWith(`Retrying beginTransaction.`)
+    })
+
+    it('should log the second error', () => {
+      expect(helpers.logError).to.be.calledWith(`Error running the beginTransaction query: Error: some error`)
+    })
+
+    it('should return null', () => {
+      expect(pgClient).to.be.null
+    })
+  })
+})
```

Comment:
I think it would be valuable to write some test cases where the return value of `beginTransaction` is used as a condition for running some subsequent code that is stubbed. That way, you can check that subsequent code is not run in the event of a deadlock error. This could be added as an additional `it` statement within the describes you've already implemented.

Reply:
Hmm I'm not sure I understand what you mean, do you mind clarifying? Is it that, if there's a deadlock you want to have a test that `return null` doesn't happen on line 85?

Reply:
I mean in the sense that the reason we rolled back to v10.0.0 was because a call to `beginTransaction` that needed to retry had exited early, and continued onto later code, while the retry call was still running. If you could replicate that scenario in a test case in that other code waits for `beginTransaction` to complete, and you test for whether or not that code was run, that would be good to observe.

Reply:
Thanks for clarifying Micah! I see what you mean but I think what you're describing is already being tested (correct me if I'm wrong). 

Currently in `db.js` the only line after the retry is `return null` on line 85. So if there was a deadlock, it would go to the catch block from line line 70-82, on line 82 the retry would get called. If the code wasn't working as expected and continued on (in the way that caused the bug) then it would hit line 85 and return `null`. This is case is already tested in the first describe block on line 55-57 in `beginTransactionTest.js`:
```
    it('should return a valid pgClient', () => {
      expect(pgClient).to.be.equal(clientStub)
    })
```
If the code continued on and hit line 85 (returning null) this test would not pass.

Reply:
Cool! I didn't notice that, but I think that's sufficient. Thanks!

Reply:
Thanks Micah!